### PR TITLE
Annotation updates

### PR DESCRIPTION
**Pull request by Ninjani**
TODO:
- Crystal contacts
  - [x] add fraction of crystal contacts as part of validation criteria
  - [x] add number of atoms to `symmetry_mate_contacts` instead of residues
  - [x] add tests
- Binding affinity
  - [x] drop IC50
  - [x] add to split as prioritization
- Split criteria
  - [x] add min/max number of pocket residues and interactions for test

### CODE REVIEW

**Comment by OleinikovasV** [comment]
> TODO:
> 
> * Crystal contacts
>   
>   * [ ]  add fraction of crystal contacts as part of validation criteria
>   * [ ]  add number of atoms to `symmetry_mate_contacts` instead of residues
>   * [ ]  add tests
> * Binding affinity
>   
>   * [ ]  drop IC50
>   * [ ]  add to split as prioritization
> * Split criteria
>   
>   * [ ]  add min/max number of pocket residues and interactions for test


- [ ] remove `system.cif` 
- [ ] add unique plip counter

File: src/plinder/data/utils/annotations/interaction_utils.py [hunk 1]
```diff
@@ -75,6 +76,37 @@ def read_mmcif_file(mmcif_filename: Path) -> PDBxFile:
...
+def get_symmetry_mate_contacts(
+    mmcif_filename: Path, contact_threshold: float = 5.0
+) -> dict[tuple[str, int], dict[tuple[str, int], set[int]]]:
+    """
+    Get all contacts within a given threshold between residues which are not in the same asymmetric unit (symmetry mates)
+    """
+    cif = gemmi.read_structure(mmcif_filename.__str__(), merge_chain_parts=False)
+    cif.setup_entities()
+    ns = gemmi.NeighborSearch(cif[0], cif.cell, contact_threshold).populate(
+        include_h=False
```
Comment (Ninjani) [inline src/plinder/data/utils/annotations/interaction_utils.py:88]: @OleinikovasV okay to ignore crystal contacts with hydrogens?

File: src/plinder/data/splits.py [hunk 2]
```diff
@@ -53,43 +31,57 @@ class GraphConfig:
...
+    # max number of representatives from each test cluster
     num_test_representatives: int = 3
     # test should not be singletons
     min_test_cluster_size: int = 5
-    # test/val should not be in too big communities or cause too many train cases to be removed
+    # test should not be in too big communities or cause too many train cases to be removed
     max_test_leakage_count: int = 300
-    mms_unique_quality_count: int = 3
+    # test should not have too few or too many interactions
+    min_max_test_pli: tuple[int, int] = (2, 50)
```
Comment (Ninjani) [inline src/plinder/data/splits.py:59]: @OleinikovasV we need to choose some values for min/max_test/val_pocket/pli. Suggest checking the cosmograph to see the distribution

File: src/plinder/data/splits.py [hunk 3]
```diff
@@ -612,14 +560,18 @@ def prioritize_test_sample(
...
+        & (test_data["system_num_interactions"] >= cfg.split.min_max_test_pli[0])
+        & (test_data["system_num_interactions"] <= cfg.split.min_max_test_pli[1])
     ]
     LOG.info(
         f"found {len(set(test_data['system_id']))} test systems after removing by second pass neighbor leakage count"
     )
     test_data = (
         test_data.sort_values(
-            ["system_has_mms", "has_apo", "leakage_count"],
+            ["system_has_mms", "has_apo", "has_binding_affinity", "leakage_count"],
```
Comment (Ninjani) [inline src/plinder/data/splits.py:574]: @OleinikovasV correct priority?

File: src/plinder/data/pipeline/transform.py [hunk 4]
```diff
@@ -133,14 +133,15 @@ def calc_pchembl(affinity: float) -> Any:
...
-        "IC50 (nM)",
+        # "IC50 (nM)",
         "Kd (nM)",
         "EC50 (nM)",
     ]
     df = pd.read_csv(raw_affinity_path, sep="\t", usecols=cols, low_memory=False)
 
     df["pchembl"] = (
-        df[["Ki (nM)", "IC50 (nM)", "Kd (nM)"]]
+        # df[["Ki (nM)", "IC50 (nM)", "Kd (nM)"]]
```
Comment (Ninjani) [inline src/plinder/data/pipeline/transform.py:143]: @yusuf1759 is this the right way to drop IC50 from bindingdb? Anything else needed for papyrus?

**Comment by Ninjani** [comment]
>     * [ ]  remove `system.cif`
> 
>     * [ ]  add unique plip counter

Was trying these out, and both not immediately trivial:

system.cif removal itself is easy but needs changing `final_structure_qc` to load receptor and ligand separately. @yusuf1759 would need changing all the `complex_path`s.


similarity scoring for unique pli, what's the strategy here - do we count each type of interaction once for each residue or just the residues themselves? e.g 
```
system_1: {res1: [hbond, hbond, saltbridge], res2: [hydrophobic, hbond]}
system_2: {res1: [hbond, saltbridge, saltbridge, hydrophobic], res2: [hbond], res3: [hydrophobic]}
```
do we want similarity to be 2/2 for system_1 vs system_2 since they both share res1 and res2 as interacting residues (irrespective of actual interactions), or we want to compare system_1_res1: {hbond, saltbridge} vs system_2_res1: {hbond, saltbridge, hydrophobic} (i.e taking the set and ignoring the count)? Or both with former being `pocket_interacting_qcov` and latter being `pli_unique_qcov`?

File: src/plinder/data/utils/annotations/get_ligand_validation.py [hunk 5]
```diff
@@ -49,51 +54,73 @@ class SystemValidationThresholds(BaseModel):
...
+                min_average_rscc=0.8,
+                max_average_rsr=0.3,
+                max_percent_outliers_clashes=100.0,
+                min_percent_rsr_under_threshold=0.0,
+                min_percent_rscc_over_threshold=0.0,
+                min_percent_occupancy_over_threshold=0.0,
             ),
         }
     )
+    max_fraction_atoms_with_crystal_contacts: float = 0.0
```
Comment (Ninjani) [inline src/plinder/data/utils/annotations/get_ligand_validation.py:88]: @OleinikovasV added fraction threshold here, should we try with strictest value (0) and see how many we lose? Does mean that if we change it we'll have to redo the system_pass_validation_criteria calculation on the index file - @tjduigna what's the best strategy here, we could move this to core instead of here to account for the fact that changing these thresholds does not actually need annotation generation rerun?

File: src/plinder/data/splits.py (hunk 3 above)
Comment (OleinikovasV) [reply src/plinder/data/splits.py]: agree!

File: src/plinder/data/utils/annotations/interaction_utils.py (hunk 1 above)
Comment (OleinikovasV) [reply src/plinder/data/utils/annotations/interaction_utils.py:88]: yes, this makes sense to me only to consider heavy atoms for a given cutoff

File: src/plinder/data/splits.py (hunk 2 above)
Comment (OleinikovasV) [reply src/plinder/data/splits.py:59]: Cosmograph suggest there would be 371 418 systems within this range - which is a fairly permissive number; the key part is to avoid cases where only one interaction is counted.

I am a bit borderline on 2. Raising this to 3,50 - would drop 33 007 systems.

This might work still better with out 50% PLI splitting - avoiding the case of matching one of two PLIs, and counting that as leak.

File: src/plinder/data/pipeline/transform.py (hunk 4 above)
Comment (OleinikovasV) [reply src/plinder/data/pipeline/transform.py:143]: Seems like papyrus would include EC50 and IC50 data 😱

image

Although with some curation modification - but it is still 😱 according to this
https://pubs.acs.org/doi/10.1021/acs.jcim.4c00049

Will need to review this to see if this makes sense, maybe we should just map to curated datasets from polaris?
https://polarishub.io/datasets?accessTypes=public

File: src/plinder/data/utils/annotations/get_ligand_validation.py (hunk 5 above)
Comment (OleinikovasV) [reply src/plinder/data/utils/annotations/get_ligand_validation.py:88]: I would change this to `max_atoms_with_crystal_contacts: int = 0` but this would be only for ligand.
Same would be for this `min_num_resolved_heavy_atoms: int = 2` - only for ligand criteria, I suppose.

File: src/plinder/data/pipeline/transform.py (hunk 4 above)
Comment (OleinikovasV) [reply src/plinder/data/pipeline/transform.py:143]: We could also keep this `bindingDB` so far, comment out `papyrus` to avoid inconsistency issues, and see what we do with this once we have a full strategy on how to potentially integrate `polaris` or `bigbind` sets.

Comment (OleinikovasV) [reply src/plinder/data/pipeline/transform.py:143]: Commented out `papyrus` in [f022c83](https://github.com/plinder-org/plinder/pull/14/commits/f022c838b47003899d6c1a6e3465c3a84ad1a964) - leaves only `bindingDB` with "Ki (nM)" and "Kd (nM)" data

Comment (OleinikovasV) [reply src/plinder/data/pipeline/transform.py:143]: [a385a41](https://github.com/plinder-org/plinder/pull/14/commits/a385a41daf3c37a5adb09eb0b6a81d215cc0ed8d) - changed to (3, 50)

File: src/plinder/data/splits.py (hunk 2 above)
Comment (OleinikovasV) [reply src/plinder/data/splits.py]: [a385a41](https://github.com/plinder-org/plinder/pull/14/commits/a385a41daf3c37a5adb09eb0b6a81d215cc0ed8d) - changed to (3, 50)

**Comment by OleinikovasV** [comment]
> ```
> system_1: {res1: [hbond, hbond, saltbridge], res2: [hydrophobic, hbond]}
> system_2: {res1: [hbond, saltbridge, saltbridge, hydrophobic], res2: [hbond], res3: [hydrophobic]}
> ```
> 
> do we want similarity to be 2/2 for system_1 vs system_2 since they both share res1 and res2 as interacting residues (irrespective of actual interactions), or we want to compare system_1_res1: {hbond, saltbridge} vs system_2_res1: {hbond, saltbridge, hydrophobic} (i.e taking the set and ignoring the count)? Or both with former being `pocket_interacting_qcov` and latter being `pli_unique_qcov`?

@Ninjani, this is a good question. I like `pocket_interacting_qcov` for matched residues that are interacting - but as long as there is at least one matched interaction, eg. if there is a 'hydrophobic' vs 'salt bridge' - I do not consider that it would be reasonable to match them. These would be already matched by the "neighbouring" residues metric of `pocket_qcov`, so, matching interaction type seems more reasonable to me.

The `pli_unique_qcov` would be the same as `pli_qcov` but only counting each unique match once.
image

**Comment by Ninjani** [comment]
@OleinikovasV I've implemented `pli_unique_qcov` already, would consider deferring `pocket_interacting_qcov` and the removal of `system.cif` to a later stage so we can do the rerun and have the test set asap.

File: src/plinder/data/utils/annotations/get_ligand_validation.py (hunk 5 above)
Comment (Ninjani) [reply src/plinder/data/utils/annotations/get_ligand_validation.py:88]: We should move these criteria to `splits` and change as necessary after annotation rerun as well

Comment (tjduigna) [reply src/plinder/data/utils/annotations/get_ligand_validation.py:88]: We could do something like:
```python

entry = Entry.from_json(stored_entry_json)
entry.set_validation(val_file, cif_file, any_thresholds)
# pseudocode -> entry.to_json()
```
which avoids annotation generation itself by re-using what we already have, but allows us to pass new thresholds.

Operationally, this would additionally require updating the zipped entries and re-generating the index. Today there is no "update annotations" stage, but it would be relatively straightforward to write. `structure_qc` could be extended to only re-pack the `entries` zips and `make_plindex` is something that should have its own stage but currently runs inside of `make_mmp_index`.

File: src/plinder/data/utils/annotations/get_similarity_scores.py [hunk 6]
```diff
@@ -56,6 +53,7 @@
     "pocket_fident",
     "pocket_fident_qcov",
     "pli_qcov",
+    "pli_unique_qcov",
```
Comment (tjduigna) [inline src/plinder/data/utils/annotations/get_similarity_scores.py:56]: Is there a goal to re-run scoring for this particular metric?

File: src/plinder/data/utils/annotations/get_ligand_validation.py (hunk 5 above)
Comment (Ninjani) [reply src/plinder/data/utils/annotations/get_ligand_validation.py:88]: Could also be that we use the `pass_validation_criteria` columns only in the plindex and not in the entry jsons? Depends on what would mesh better with the dataloader

**Review by tjduigna — APPROVED** [review]

File: src/plinder/data/utils/annotations/get_ligand_validation.py (hunk 5 above)
Comment (tjduigna) [reply src/plinder/data/utils/annotations/get_ligand_validation.py:88]: Yeah we could opt not to save the `pass_validation_criteria` in the entry jsons and only evaluate them in `to_df` on index creation. Then we'd only be storing the data in one place, and we'd simply have to call `set_validation` in `structure_qc` prior to `Entry.to_df`. It would make sense to surface the validation configuration the same way we have it as for other configurable parts of the system.
